odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2372 @@
+        """
+        self.data.real[:] = newreal
+
+    @property
+    def imag(self):
+        """Imaginary part of this tensor.
+
+        Returns
+        -------
+        imag : `CupyTensor`
+            The imaginary part of this tensor as an element of an `rn` space.
+        """
+        if self.space.is_real:
+            return self.space.zero()
+        else:
+            return self.space.real_space.element(self.data.imag)
+
+    @imag.setter
+    def imag(self, newimag):
+        """Setter for the imaginary part.
+
+        This method is invoked by ``tensor.imag = other``.
+
+        Parameters
+        ----------
+        newimag : `array-like` or scalar
+            The new imaginary part for this tensor.
+        """
+        self.data.imag[:] = newimag
+
+    def conj(self, out=None):
+        """Complex conjugate of this tensor.
+
+        Parameters
+        ----------
+        out : `CupyTensor`, optional
+            Tensor to which the complex conjugate is written.
+            Must be an element of this tensor's space.
+
+        Returns
+        -------
+        out : `CupyTensor`
+            The complex conjugate tensor. If ``out`` was provided,
+            the returned object is a reference to it.
+        """
+        if out is None:
+            if self.space.is_real:
+                return self.copy()
+            else:
+                return self.space.element(self.data.conj())
+        else:
+            if self.space.is_real:
+                self.assign(out)
+            else:
+                # In-place not available as it seems
+                out[:] = self.data.conj()
+            return out
+
+    def __ipow__(self, other):
+        """Return ``self **= other``."""
+        try:
+            if other == int(other):
+                return super(CupyTensorSpace, self).__ipow__(other)
+        except TypeError:
+            pass
+
+        self.ufuncs.power(self.data, other, out=self.data)
+        return self
+
+
+# --- Weightings --- #
+
+
+def _weighting(weights, exponent):
+    """Return a weighting whose type is inferred from the arguments."""
+    if np.isscalar(weights):
+        weighting = CupyTensorSpaceConstWeighting(weights, exponent=exponent)
+    else:
+        # TODO: sequence of 1D array-likes, see
+        # https://github.com/odlgroup/odl/pull/1238
+        weights = cupy.array(weights, copy=False)
+        weighting = CupyTensorSpaceArrayWeighting(weights, exponent=exponent)
+    return weighting
+
+
+# Kernels for space functions
+#
+# T = generic type
+# R = real floating point type, usually for norm or dist output
+# W = real type for weights, can be floating point or integer
+#
+# Note: the kernels with an output type that does not also occur as an input
+# type must be called with output argument since the output type cannot be
+# inferred. The ouptut array must have shape `()` for full reduction.
+
+
+if CUPY_AVAILABLE:
+    dotw = cupy.ReductionKernel(in_params='T x, T y, W w',
```

Comment:
Are these compiled on the fly later? We don't want to create startup latency

Reply:
Yes, as I wrote somewhere else, the import time of this module is practically identical to the import time of `cupy`, around 40 ms. So these guys don't add overhead at creation time.